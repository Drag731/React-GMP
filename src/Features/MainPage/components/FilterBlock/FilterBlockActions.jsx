export const SEARCH_MOVIE = 'SEARCH_MOVIE';
export const SET_SORT_BY = 'SET_SORT_BY';
export const SET_SEARCH_BY = 'SET_SEARCH_BY';
export const SEARCH_BUTTON = 'SEARCH_BUTTON';

export const handleSearch = (search) => ({
    type: SEARCH_MOVIE,
    search
});

export const handleSearchButton = () => ({
    type: SEARCH_BUTTON,
});

export const setSearchBy = (payload) => ({
    type: SET_SEARCH_BY,
    payload
});

export const setSortBy = (payload) => ({
    type: SET_SORT_BY,
    payload
});