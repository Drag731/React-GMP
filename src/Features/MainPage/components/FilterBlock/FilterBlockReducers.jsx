import {
    SEARCH_MOVIE,
    SET_SORT_BY,
    SET_SEARCH_BY,
} from './FilterBlockActions';

import {
    DEFAULT_SORT,
    DEFAULT_SEARCH
} from '../../../../constants/constants'

const initialState = {
    search: '',
    sortBy: DEFAULT_SORT,
    searchBy: DEFAULT_SEARCH
};

const FilterReducer = (state = initialState, action) => {
    switch (action.type) {

        case SET_SORT_BY: {
            return {
                ...state,
                isLoadingMovies: true,
                sortBy: action.payload
            };
        }

        case SET_SEARCH_BY: {
            return {
                ...state,
                searchBy: action.payload
            };
        }

        case SEARCH_MOVIE: {
            return {
                ...state,
                search: action.search
            };
        }

        default: {
            return state;
        }
    }
};

export const getQuery = state => {
    let {search, searchBy, sortBy} = state.filter;
    if (searchBy === 'genres' && search !== '') {
        search = search[0].toUpperCase() + search.slice(1);
    }
    return `?search=${search}&searchBy=${searchBy}&sortBy=${sortBy}&sortOrder=desc`
};

export default FilterReducer;