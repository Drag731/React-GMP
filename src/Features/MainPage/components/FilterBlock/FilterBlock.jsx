import React, { Component } from 'react';
import { connect } from 'react-redux';

import ResultBlock from '../ResultBlock/ResultBlock';
import HeaderBlock from '../HeaderBlock/HeaderBlock';

import {
    handleSearch,
    handleSearchButton,
    setSearchBy,
    setSortBy,
    handleEnterPress
} from './FilterBlockActions';

import { getQuery } from './FilterBlockReducers';

import { receiveMoviesDB } from '../../MainPageActions';
import { getTotal } from '../../MainPageReducers';

const mapStateToProps = state => ({
    total: getTotal(state),
    query: getQuery(state),
    search: state.filter.search,
    searchBy: state.filter.searchBy,
    sortBy: state.filter.sortBy,
});

const mapDispatchToProps = {
    receiveMoviesDB: (q) => receiveMoviesDB(q),
    handleSearch: (search) => handleSearch(search),
    setSearchBy: payload => setSearchBy(payload),
    setSortBy: payload => setSortBy(payload),
    handleSearchButton: () => handleSearchButton(),
    handleEnterPress: () => handleEnterPress()
};

class FilterBlock extends React.Component {
    constructor(props) {
        super(props);

        this.handlePop = this.handlePop.bind(this);
    }

    componentDidMount() {
        if (this.props.location.search !== "") {
            this.transitionToURL();
        }
        this.props.receiveMoviesDB(this.props.location.search);
    }

    componentWillReceiveProps(nextProps) {
        if (this.props.sortBy !== nextProps.sortBy) {
            this.props.receiveMoviesDB(nextProps.query);
            this.setUrlParams(nextProps);
        }
    }

    componentWillMount() {
        addEventListener("popstate", this.handlePop)
    }

    componentWillUnmount() {
        removeEventListener("popstate", this.handlePop)
    }

    handlePop() {
        if (this.props.location.search !== "") {
            this.transitionToURL();
        }
        this.props.receiveMoviesDB(this.props.location.search);
    };

    handleSearch = (search) => { this.props.handleSearch(search)};

    handleSearchButton = () => {
        this.props.handleSearchButton();
        this.props.receiveMoviesDB(this.props.query);
        this.setUrlParams();
    };

    setUrlParams = (props = this.props) => {
        props.history.push({
            pathname: props.location.pathname,
            search: props.query
        });
    };

    handleTitleSearch = () => { this.props.setSearchBy('title')};

    handleGenresSearch = () => { this.props.setSearchBy('genres')};

    handleReleaseDateSort = () => { this.props.setSortBy('release_date') };

    handleRatingSort = () => { this.props.setSortBy('vote_average') };

    handleEnterPress = (e) => {
        if(e.keyCode === 13) {
            this.props.handleEnterPress();
            this.props.receiveMoviesDB(this.props.query);
        }
    };

    transitionToURL = () => {
        const params = new URLSearchParams(this.props.location.search);
        this.props.setSearchBy(params.get('searchBy'));
        this.props.setSortBy(params.get('sortBy'));
        this.props.handleSearch(params.get('search'));
    };

    render() {
        const { search, total, searchBy, sortBy } = this.props;

        return (
            <React.Fragment>
                <HeaderBlock
                    handleSearch={this.handleSearch}
                    handleSearchButton={this.handleSearchButton}
                    handleTitleSearch={this.handleTitleSearch}
                    handleGenresSearch={this.handleGenresSearch}
                    handleEnterPress={this.handleEnterPress}
                    search={search}
                    searchBy={searchBy}
                />
                <ResultBlock
                    total={total}
                    handleReleaseDateSort={this.handleReleaseDateSort}
                    handleRatingSort={this.handleRatingSort}
                    sortBy={sortBy}
                />
            </React.Fragment>
        )
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(FilterBlock);
export const FilterBlockTest = FilterBlock;
