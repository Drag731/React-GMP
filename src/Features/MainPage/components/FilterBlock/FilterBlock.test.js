import React from 'react';
import renderer from 'react-test-renderer';
import {shallow, mount, configure} from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

import { FilterBlockTest } from './FilterBlock'

configure({ adapter: new Adapter() });

describe('<FilterBlockTest  />', () => {
    const mockProps = {
        search: '',
        sortBy: '',
        searchBy: 'title',
        total: 3000,
        query: '',
        handleSearch: () => {},
        receiveMoviesDB: () => {},
        setSearchBy: () => {},
        setSortBy: () => {},
        handleSearchButton: () => {},
        handleEnterPress: () => {},
    };

    it('Renders FilterBlockTest component', () => {
        const tree = shallow(
            < FilterBlockTest {...mockProps} />
        );
        expect(toJson(tree)).toMatchSnapshot();
    });
    it('Renders FilterBlockTest component snapshot2', () => {
        const tree = mount(
            < FilterBlockTest {...mockProps} />
        );
        expect(toJson(tree)).toMatchSnapshot();
    });
    it('FilterBlockTest gets new props', () => {
        const sortBy1 = 'date';
        const sortBy2 = 'rating';
        const receiveMoviesDB = jest.fn();
        const tree = mount(
            <FilterBlockTest
                {...mockProps}
                sortBy={sortBy1}
                receiveMoviesDB={receiveMoviesDB}
            />
        );

        expect(tree.prop('sortBy')).toEqual(sortBy1);
        tree.setProps({sortBy: sortBy2});
        expect(tree.prop('sortBy')).toEqual(sortBy2);
        expect(receiveMoviesDB).toHaveBeenCalled();
    });
    it('should call handleRatingSort on click', () => {
        const tree = mount(
            <FilterBlockTest {...mockProps} sortBy="release_date"/>
        );
        const spy = jest.spyOn(tree.instance(), 'handleRatingSort');
        tree.update();
        tree.find('.b-result__sort-item').first().simulate('click');
        expect(spy).toHaveBeenCalled();
    });
});
