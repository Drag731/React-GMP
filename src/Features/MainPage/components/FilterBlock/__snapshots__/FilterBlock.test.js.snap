// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterBlockTest  /> Renders FilterBlockTest component 1`] = `
<React.Fragment>
  <HeaderBlock
    handleEnterPress={[Function]}
    handleGenresSearch={[Function]}
    handleSearch={[Function]}
    handleSearchButton={[Function]}
    handleTitleSearch={[Function]}
    search=""
    searchBy="title"
  />
  <ResultBlock
    handleRatingSort={[Function]}
    handleReleaseDateSort={[Function]}
    sortBy=""
    total={3000}
  />
</React.Fragment>
`;

exports[`<FilterBlockTest  /> Renders FilterBlockTest component snapshot2 1`] = `
<FilterBlock
  handleEnterPress={[MockFunction]}
  handleSearch={
    [MockFunction] {
      "calls": Array [
        Array [
          null,
        ],
        Array [
          null,
        ],
      ],
    }
  }
  handleSearchButton={[MockFunction]}
  location={
    Object {
      "search": "a",
    }
  }
  query=""
  receiveMoviesDB={
    [MockFunction] {
      "calls": Array [
        Array [
          "a",
        ],
        Array [
          "a",
        ],
      ],
    }
  }
  search=""
  searchBy="title"
  setSearchBy={
    [MockFunction] {
      "calls": Array [
        Array [
          null,
        ],
        Array [
          null,
        ],
      ],
    }
  }
  setSortBy={
    [MockFunction] {
      "calls": Array [
        Array [
          null,
        ],
        Array [
          null,
        ],
      ],
    }
  }
  sortBy=""
  total={3000}
>
  <HeaderBlock
    handleEnterPress={[Function]}
    handleGenresSearch={[Function]}
    handleSearch={[Function]}
    handleSearchButton={[Function]}
    handleTitleSearch={[Function]}
    search=""
    searchBy="title"
  >
    <div
      className="main-page__header b-header"
    >
      <HeaderLogo
        description={false}
      >
        <div
          className="b-header__logo-wrap"
        >
          <h1
            className="b-header__logo-text"
          >
            Filmzilla
          </h1>
        </div>
      </HeaderLogo>
      <SearchBlock
        handleEnterPress={[Function]}
        handleSearch={[Function]}
        search=""
      >
        <div
          className="b-header__search"
        >
          <span
            className="b-header__search-title"
          >
            find your movie
          </span>
          <input
            className="b-header__search-input"
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="search"
            type="search"
            value=""
          />
        </div>
      </SearchBlock>
      <ChoiceBlock
        handleGenresSearch={[Function]}
        handleSearchButton={[Function]}
        handleTitleSearch={[Function]}
        search=""
        searchBy="title"
      >
        <div
          className="b-header__choice choice-flex"
        >
          <div
            className="choice-flex__item"
          >
            <span
              className="b-header__choice-text"
            >
              search by
            </span>
            <Button
              className="b-header__button_active"
              handler={[Function]}
            >
              <button
                className="b-header__button_active"
                onClick={[Function]}
              >
                title
              </button>
            </Button>
            <Button
              className="b-header__button"
              handler={[Function]}
            >
              <button
                className="b-header__button"
                onClick={[Function]}
              >
                genres
              </button>
            </Button>
          </div>
          <div
            className="choice-flex__item"
          >
            <Button
              handler={[Function]}
              style={
                Object {
                  "background": "#ccc",
                  "pointerEvents": "none",
                }
              }
            >
              <button
                className=""
                onClick={[Function]}
                style={
                  Object {
                    "background": "#ccc",
                    "pointerEvents": "none",
                  }
                }
              >
                search
              </button>
            </Button>
          </div>
        </div>
      </ChoiceBlock>
    </div>
  </HeaderBlock>
  <ResultBlock
    handleRatingSort={[Function]}
    handleReleaseDateSort={[Function]}
    sortBy=""
    total={3000}
  >
    <div
      className="main-page__result b-result"
    >
      <div
        className="b-result__text-wrap"
      >
        <span
          className="b-result__text"
        >
          3000
           movies found
        </span>
      </div>
      <div
        className="b-result__sort-wrap"
      >
        <span
          className="b-result__sort-text"
        >
          Sort by:
        </span>
        <span
          className="b-result__sort-item"
          onClick={[Function]}
        >
          release date
        </span>
        <span
          className="b-result__sort-item"
          onClick={[Function]}
        >
          rating
        </span>
      </div>
    </div>
  </ResultBlock>
</FilterBlock>
`;
